### PR TITLE
Fixed #33405, Refs #7177 -- Clarified docs for filter escapejs regarding safe and usafe usages.

### PR DESCRIPTION
**Pull request by jribbens**
Fixes ticket-33405

### CODE REVIEW

File: docs/ref/templates/builtins.txt
Comment (carltongibson) [inline docs/ref/templates/builtins.txt:1647]: I don't think this is right. 🤔

ticket-27900 made escapejs escape backticks to support usage in ES6 template literals. 6ae1b04fb584db0fdb22b8e287784c4ed3ac62ac

Also `safe` may not be right... i.e. is that not what's causing the confusion here? It means you won't get a syntax error using the DTL to interpolate, but it doesn't protect against XSS &co, which is what I guess _safe_ might be taken for. 

Maybe deprecating `escapejs`... — I suspect folks might complain 🤔

Comment (jribbens) [reply docs/ref/templates/builtins.txt:1647]: > I don't think this is right. 🤔
> 
> [ticket-27900](https://code.djangoproject.com/ticket/27900) made escapejs escape backticks to support usage in ES6 template literals. https://github.com/django/django/commit/6ae1b04fb584db0fdb22b8e287784c4ed3ac62ac

Eh? It was you yourself who pointed me to ticket-29055 which explains how escapejs does not work with ES6 template literals.

> Also safe may not be right... i.e. is that not what's causing the confusion here? It means you won't get a syntax error using the DTL to interpolate, but it doesn't protect against XSS &co, which is what I guess safe might be taken for.

Yes that is what "safe" means and how I am using it. Either the filter makes the content "safe" or it is useless for almost any purpose. escapejs does protect against XSS.

> Maybe deprecating escapejs... — I suspect folks might complain 🤔

I'm trying to do the opposite - to rejuvenate this useful feature that has languished unnecessarily due to its documentation being incomprehensible / incorrect.

Comment (carltongibson) [reply docs/ref/templates/builtins.txt:1647]: > Yes that is what "safe" means and how I am using it. 

OK, but if that's the meaning of "safe" then there's no usage of `escapejs` which is _safe_. It simply doesn't do that. AFAICS the only use of `escapejs` is to allow you to interpolate values from DTL template variables without needing to worry about JS syntax errors, when they contain `'` for example. 

It's strikes me the existing is quite clear on this: 

> This does *not* make the string safe for use in HTML or JavaScript template literals, but does protect
 you from syntax errors when using templates to generate JavaScript/JSON.

I kind of (still) think we should close this as `wontfix`. 

This is a difficult problem. The `json_script` tag is the way to go. https://docs.djangoproject.com/en/4.0/ref/templates/builtins/

Perhaps @adamchainz can comment, having blogged on this topic? 🙏

Comment (adamchainz) [reply docs/ref/templates/builtins.txt:1647]: I did indeed only look at as an afterthought. I don't write inline `<script>` tags ever, because they are inefficient and disallowed with a strong CSP.

The way I understood `escapejs` as being safe is for templated whole javascript files, i.e. where the template is the JS file. This may indeed be a mismatch between the docs and what the code does, as the code does escape `<` & co., so it looks like it could be safe for templating inline `<script>`s.

I'm not sure what the previous authors of the docs meant by "not safe for use in HTML". Perhaps there are cases I haven't thought of. It would be better to contact them to help with clarification.

Comment (carltongibson) [reply docs/ref/templates/builtins.txt:1647]: `escapejs` was introduced 14 years ago for ticket-4131. 

The [original diff](https://github.com/django/django/commit/b65fce659502ac5e211d13fbd5435e1ff6c703d2#diff-d247cb923bcb90ec15cafc2d94d1ba7a1bf7da7a89175902adc7431f5475a9dcR1312-R1314) has that it is **not** _safe_ for use in HTML, just to avoid syntax errors. 

> This does *not* make the string safe for use in HTML, but does protect you from syntax errors when using
 templates to generate JavaScript/JSON.

The ticket discusses that it doesn't protect correctly against XSS — though **some** protection is granted — hence the disclaimer.

Comment (jribbens) [reply docs/ref/templates/builtins.txt:1647]: Well I think the mystery is solved then. In the original diff you linked where escapejs was introduced, the way it worked was completely different from the way it works now, and probably was indeed not safe for HTML.

But in ticket-7177 the code was updated to change the way it works, the idea being to make it so that it _is_ safe for HTML. However the patch that implemented this, 8c4a525871df19163d5bfdf5939eff33b544c2e2, forgot to update the documentation.

So the situation is: the documentation was written for the original version of escapejs, but since then multiple changes have been made to the code which have caused the documentation and the implementation to diverge. So my PR is simply updating the documentation to reflect the code as it exists now.

Comment (carltongibson) [reply docs/ref/templates/builtins.txt:1647]: OK, I'll ask the security team to look.

**Comment by phith0n** [comment]
> This makes the string safe for use in HTML and JavaScript or JSON string literals

I don't think it's OK. Not just ES6 template literals, but also in javascript URL protocol. For example:

```html
<a href="javascript:console.log('{{ request.GET.q | escapejs }}')">click me</a>
```

Although the input is in the "JavaScript string literals", it still can be attacked by this payload:

```
?q=%27%2Balert%281%29%2B%27
```

**Comment by adamchainz** [comment]
@phith0n I cannot reproduce your example, it amounts to:

```
In [1]: from django.template import Template, Context

In [2]: t = Template("""<a href="javascript:console.log('{{ q | escapejs }}')">click me</a>""")

In [3]: print(t.render(Context({"q": "'+alert(1)+'"})))
<a href="javascript:console.log('\u0027+alert(1)+\u0027')">click me</a>
```

`escapejs` clearly maps the single quote to `\u0027`, as can be seen in the source:

https://github.com/django/django/blob/d19a53d8e3640e8f937048bf7826c7e2229a9665/django/utils/html.py#L30-L43

**Comment by phith0n** [comment]
@adamchainz Hi, try this one:

```
In [1]: from django.template import Template, Context

In [2]: t = Template("""<a href="javascript:console.log('{{ q | escapejs }}')">click me</a>""")

In [3]: print(t.render(Context({"q": "%27%2Balert%281%29%2B%27"})))
<a href="javascript:console.log('%27%2Balert%281%29%2B%27')">click me</a>
```

`alert(1)` will be execute when you click the link: <https://codepen.io/phithon/pen/MWQRZqO>

**Comment by phith0n** [comment]
> > @adamchainz Hi, try this one:
> > ```
> > In [2]: t = Template("""<a href="javascript:console.log('{{ q | escapejs }}')">click me</a>""")
> > ```
> 
> But that's due to a bug in your template, not a problem with `escapejs`. `q` is embedded within JavaScript code, which `escapejs` has correctly dealt with, but then that is itself embedded within a URL, and you've forgotten to escape around that. The correct template would be:
> 
> ```
> <a href="javascript:console.log('{{ q | escapejs | urlencode }}')">
> ```

There is no problem with `escapejs`, but it can mislead developers to trust the `escapejs` if you say *"this makes the string safe
for use in HTML and JavaScript or JSON string literals"*.

**Comment by jribbens** [comment]
> There is no problem with `escapejs`, but it can mislead developers to trust the `escapejs` if you say _"this makes the string safe for use in HTML and JavaScript or JSON string literals"_.

But none of those things are "javascript: URLs". It doesn't say either that you shouldn't do `eval({{ q | escapejs}})`, but you clearly shouldn't and it's your own fault if you do.

**Comment by phith0n** [comment]
> > There is no problem with `escapejs`, but it can mislead developers to trust the `escapejs` if you say _"this makes the string safe for use in HTML and JavaScript or JSON string literals"_.
> 
> But none of those things are "javascript: URLs". It doesn't say either that you shouldn't do `eval({{ q | escapejs}})`, but you clearly shouldn't and it's your own fault if you do.

I don't think `javascript:console.log('{{ q | escapejs }}')` is equal to `eval({{ q | escapejs}})`. The previous one, is put in JavaScript string literals (although it also in the URL context), not all the developer knows that they have to quote it.

Due to the "secure as default" philosophy, I think it's not a good idea to update the document. Another way is to add `%` to the  escape list in `escapejs`:

```python
_js_escapes = { 
     ord("\\"): "\\u005C", 
     ord("'"): "\\u0027", 
     ord('"'): "\\u0022", 
     ord(">"): "\\u003E", 
     ord("<"): "\\u003C", 
     ord("&"): "\\u0026", 
     ord("="): "\\u003D", 
     ord("-"): "\\u002D", 
     ord(";"): "\\u003B", 
     ord("`"): "\\u0060", 
     ord("%"): "\\u0025",
     ord("\u2028"): "\\u2028", 
     ord("\u2029"): "\\u2029", 
 } 
```

**Comment by carltongibson** [comment]
To add some context, from my perspective this is still sitting here because the feedback from the security team was more or less, _"If we can phrase it correctly, ..."_ but that conceals a lot. It's not at all clear (to me at least) what we should say, for the exact cases it's OK, and how we don't encourage people into adopting a foot-gun. 

> Due to the "secure as default" philosophy, I think it's not a good idea to update the document and say it is "safe to use in HTML".

I more or less agree with this.

**Comment by jribbens** [comment]
@carltongibson What is the problem with saying it's "safe to use in HTML"? It's true, and it doesn't violate "secure by default". What scenario is being imagined whereby a problem might arise?

The documentation really can't stay as it is - it's flat-out wrong, and so confused as to be almost meaningless. What on earth would "it isn't safe but it protects you from syntax errors" be supposed to mean even if it were true? And I guess "when using templates to generate JavaScript/JSON" is supposed to mean "when using templates to generate JavaScript/JSON files", but that isn't what it says (*obviously* you're generating JavaScript/JSON, since this filter has no other purpose).

The filter output is safe to use:

* in HTML content and attribute values
* in JavaScript/JSON `'` and `"` string literals in JavaScript/JSON files
* in JavaScript/JSON `'` and `"` string literals inside HTML in-line `<script>` content and attribute values

The things that aren't safe to do include:

* using it in JavaScript ``` ` ``` template literals
* using it in a string literal but putting a `\` immediately before the variable part
* using it in a `javascript:` URL without using `urlencode`

but the latter two of these seem so contrived as to constitute writing a caveat "do not load gun, take careful aim at own foot, and pull trigger".

**Comment by carltongibson** [comment]
The problem is there's a lot of talk but not a clear demonstration that it's actually safe. 

And then there are the examples where suddenly it's not, which would need documentation. 

If don't think _"it's safe unless you do this"_ is a very good thing to be saying. People don't read the caveats, use it badly, and then say "but you said it was safe". Better the current text.

**Comment by jribbens** [comment]
But that's what I was just saying, the current text is not "better", the current text is false, and worse than that, it's gibberish. It has no discernible meaning. There is no way that leaving it unchanged is a good idea.

I don't know what you mean by "there are the examples where suddenly it's not [safe]". The "examples" (apart from backtick template strings, which according to ticket-29055 is a deliberate decision by the security list to keep as unsafe and yet the current documentation does not clearly describe as unsafe - which my patch fixes) are of the category "if the developer writes buggy code then there will be bugs". If you wanted to document every place in Django where it is possible to write buggy code then obviously the documentation would be of infinite size.

**Comment by jribbens** [comment]
> It seems pretty clear TBH, much clearer than trying to specify a exact set of **safe** usages. (Recalling that **safe** means not subject to injection attacks)

My patch specifies the safe usages, and is very clear.

I think you're failing to appreciate how the current documentation reads to someone who doesn't already know what it means. It's like someone's winning entry for an "ambiguous phrasing" competition. It says:

> Escapes characters for use in JavaScript strings.

This sentence isn't untrue, but there is vital unmentioned information here that the filter doesn't turn a Python string into a JavaScript string, it turns it into a sequence of characters that can be *inserted into* a JavaScript string - i.e. it doesn't add leading and trailing quotes. The only way to discover this is by examining the source or through experimentation. My patch fixes this by making it clear with an example.

> This does not make the string safe for use in HTML

This part is false, and incredibly confusing: "If it doesn't make it safe for HTML, why on earth does this filter exist, what is it *for*?" - bear in mind, 99% of the usage of Django templates is to create HTML files, not other sorts of files. My patch fixes this by removing the "not".

> or JavaScript template literals,

This part is extremely easy to misunderstand - it means "the ES6 JavaScript feature of template literals" but given this whole document is describing Django templates, the more obvious interpretation of it is "Django templates that output JavaScript literals". My patch fixes this by making it explicit what we're talking about.

> but does protect you from syntax errors

This part is meaningless gibberish. What bizarre meaning of the word "protect" is the documentation using if we can be simultaneously "protected" and yet also "unsafe"? Why would anyone want to generate syntactically-valid but unsafe code!? My patch fixes this by removing it.

> when using templates to generate JavaScript/JSON.

I guess this means "when generating JavaScript/JSON files" *but it doesn't say that*. Anyone looking for something to generate JavaScript/JSON that is embedded within HTML will think this part is talking about them, which it isn't. My patch fixes this by removing it.

**Comment by carltongibson** [comment]
Hi @jribbens. 

We seem to be going round in circles... 🤔

The point about the output being unsafe is (I take it) that you can't pass arbitrary data from an untrusted source without risk of an injection. The filter is still useful because I can use it with data that is trusted (that I created) to construct syntactically valid JavaScript. 

Maybe there's a case for clarification, adding the usage example, and specifying "JavaScript template literals", and so on. 

What I'm objecting to here is making the strong claim that, in fact, despite previous judgements to the contrary, `escapejs` is **safe** to use with arbitrary data from an untrusted source. I don't see either the argument or the test cases to support that. 

We frequently see security reports in areas otherwise thought safe, and am not keen to make this change unless we can show that it's correct.  Even in the simple case, is there really no injection vulnerability? Are we sure about that? (I'm not.) Then, for me, the _it's safe in this case, but not that one_ claim is just a sign that the guidance should be, as the existing, _it's not safe_, i.e. users should use `escapejs` with untrusted data.

**Comment by jribbens** [comment]
> We seem to be going round in circles... 🤔
> 
> The point about the output being unsafe is (I take it) that you can't pass arbitrary data from an untrusted source without risk of an injection. The filter is still useful because I can use it with data that is trusted (that I created) to construct syntactically valid JavaScript.
> 
> Maybe there's a case for clarification, adding the usage example, and specifying "JavaScript template literals", and so on.

Yes, I mean I don't like this solution but it is still an improvement on "change nothing", because as explained above, if this is the meaning the documentation is supposed to be conveying, it most certainly does not convey that meaning at present.

> What I'm objecting to here is making the strong claim that, in fact, despite previous judgements to the contrary, `escapejs` is **safe** to use with arbitrary data from an untrusted source. I don't see either the argument or the test cases to support that.

I don't know what you mean by "previous judgements to the contrary". As far as I can tell, no such judgements exist. What has happened is that in ticket-7177 the filter was significantly modified to make it safe, and they simply forgot to update the documentation at the same time. Nobody has ever actually decided the filter (in the version that has existed for a great many years) is unsafe.

"Test cases" to prove security obviously makes no sense, but I can provide argument as to why it is safe if you wish. It will obviously be slightly lengthy and I am aware this thread has already ceased being short some time ago, so please let me know if you think it would be helpful, rather than annoying, for me to provide a proof of safety.

I can also update the patch as suggested above by @phith0n to make it safe for the extremely obscure case of `javascript:` URLs where the developer forgets to call `urlencode`, if that would be helpful. I *could* also make it safe for use in ES6 template literals, but as mentioned in ticket-29055, it is apparently a deliberate decision by the security list that the filter be unsafe in that situation.

**Comment by carltongibson** [comment]
> so please let me know if you think it would be helpful, rather than annoying, for me to provide a proof of safety.

If you can make it plain, then yes, I'd be very happy to read it! Thanks.

**Comment by jribbens** [comment]
Ok. So, firstly, what does `escapejs` do *exactly*? We can see from [django/utils/html.py](https://github.com/django/django/blob/main/django/utils/html.py) that it transforms the following characters:
```
\x00-\x1f  \  '  "  >  <  &  =  -  ;  `  \u2028  \u2029
```
and all of them are replaced with sequences of the form `\uXXXX` where `XXXX` is 4 hexadecimal digits.

The most fundamental and important question is, of course, is the output from this safe to be passed to something that will execute it as JavaScript? We can inspect the [ECMAScript specification](https://262.ecma-international.org/#sec-literals-string-literals) to see that:

> All code points may appear literally in a string literal except for the closing quote code points, U+005C (REVERSE SOLIDUS), U+000D (CARRIAGE RETURN), and U+000A (LINE FEED). Any code points may appear in the form of an escape sequence.

We can immediately see that all of those five characters are indeed encoded by `escapejs`, and that the `\uXXXX` escapes they are replaced with are allowed and are valid instances of the JavaScript [UnicodeEscapeSequence](https://262.ecma-international.org/#prod-UnicodeEscapeSequence).

So next let's look at the [JSON grammar](https://www.json.org/json-en.html) to see if the `escapejs` output is ok there. That says we can have all characters `\u0020` - `\u10ffff` except `"` and `\`, and that the `\uXXXX` escapes are valid. `escapejs` is encoding `"`, `\` and `\x00-\x1f` so we're all good for JSON.

It is important to note that all of the above assumes that you are using the filter in the way I have documented it, i.e. one of `"{{ foo|escapejs }}"` or `'{{ foo|escapejs }}'`. If you were to omit the quotes, or if you put a partial escape sequence immediately before the inserted value, then the output can be unsafe. *The [current documentation](https://docs.djangoproject.com/en/dev/ref/templates/builtins/#escapejs) does not mention this, so my patch improves this situation.* If you think this should be made even more explicit, I can of course easily do this. But note, this is unlikely to end up as a security problem as the code will simply not work as expected even when presented with non-malicious input, so the developer is going to notice before the code is deployed.

As previously mentioned, `escapejs` output is not safe for inclusion in ES6 backtick template string literals. I think this could probably be remedied by adding `$` to the list of encoded characters, but the fact that it is unsafe is apparently a deliberate decision of the Django security list so I have not proposed to do this. My patch however does improve the situation by documenting clearly that you should not use `escapejs` in this context.

So the next question is: is the output safe for use as HTML text and HTML attribute values? The HTML specification would be an extremely verbose source to use for this, but we can presumably agree that if the `escapejs` output is at least as safe as all the rest of the output from Django templates then we're doing well.

Django delegates the escaping of HTML text and attributes to Python's [html.escape](https://github.com/python/cpython/blob/3.10/Lib/html/__init__.py) function. We can see from its source that it encodes the characters `&`, `<`, `>`, `"`, and `'`. From this we can see that (a) `escapejs` does indeed prevent any of those characters from occurring in its output, and (b) the characters that `escapejs` *adds* to its output (i.e. `\`, `u` and hexadecimal digits) are all treated as plain text by HTML. So not only is the output from `escapejs` safe to embed in HTML, but the HTML parser will leave it unchanged so the JavaScript/JSON parser will still understand it.

The above suffices to prove that it is correct to say, as my patch does, that the output of `escapejs` is "safe for use in HTML and JavaScript or JSON string literals".

The only other point mentioned is that of `javascript:` URLs. Personally I think this is not particularly relevant, because (a) the documentation does not say that the output is safe to use in this situation, either before or after my patch, and (b) the developer should know that if they're encoding any sort of value into a URL then they must call the `urlencode` filter.

However, as @phith0n mentioned, perhaps even this unwise and obscure situation could be made safe by adding `%` to the list of characters that `escapejs` encodes. We can see from the [HTML specification](https://html.spec.whatwg.org/multipage/browsing-the-web.html#javascript-protocol) that the URL is processed by [percent-decoding](https://url.spec.whatwg.org/#string-percent-decode) and then [UTF-8 decoding](https://encoding.spec.whatwg.org/#utf-8-decode) its contents. "Percent decoding", as the name suggests, only affects sequences of the form `%XX`, and as above, `escapejs` will never introduce a `%` character as part of its encoding, so making it encode `%` would indeed make this situation safe as well. I can add this to my patch if that would be thought helpful.

**Comment by carltongibson** [comment]
Hi @jribbens — sorry for the delay I had COVID. I will pick this up again to review.

**Comment by apollo13** [comment]
Hi there, I just ran over this ticket and those are my thoughts towards `escapejs`:
Independent of what the filter does, I do not want it to be used widely as it sends a bad usage signal. The main intended usage is the one that this PR adds an example for, namely:
```
    <script>
    let myValue = '{{ value|escapejs }}'
    let myValue2 = "{{ value|escapejs }}"
    </script>
```
And in my opinion supporting even that is going further than I want nowadays. I personally do not see any reason to include JS literally in HTML like this. Javascript should be loaded externally and required data for it should come via `data-*` attributes on HTML elements (or via `|json_script`).

Even if this filter is usable for JSON generation, I would not want anyone to use it for JSON. Manually constructing JSON like that sets a bad precedent imo; next thing people start constructing XML manually.

So from a security (strict CSP) and best practices POV there is no reason to have this filter. Having this filter at all encourages "bad" practices. Mind you, this filter is old, so I understand why we have it. Can we simply deprecate and remove it? ;)

In that sense I would trim & simplify the docs to say something the lines of (obviously needs some refinements but you should get the gist):
> Escapes characters for use in JavaScript strings. This makes the string safe for use in  `<script>` tags of a HTML document or in a javascript template file. Note that it does *not* make the string safe for use in "JavaScript template literals" (i.e. the JavaScript backtick syntax). Any other use is not supported, see `json_script` for an html data attributes for alternatives.

What do you think?

**Comment by apollo13** [comment]
@jribbens 

> @apollo13 I'm not sure that everyone in the world should have to write code according to your personal opinions ;-)

Ha, I do not care how people write code as long as we don't encourage bad patterns :)

> I would not want to see this filter removed.

May I ask what you are using it for and why you think there are no better alternatives?

> including when embedded in HTML.

Not sure about how I feel about that on; embedded in HMTL can mean many things. "embedded in a HTML script tag" would be clear as far as the intent is concerned.

> And keeping the example, which is vital as without it there's nothing to tell you that you have to include the quote characters yourself.

Absolutely, the example is really needed (that said one usually notices during dev that hell break loose if you don't add the quotes -- unless you test solely with numbers or so ;))

**Comment by jribbens** [comment]
@apollo13

> > I would not want to see this filter removed.
> 
> May I ask what you are using it for and why you think there are no better alternatives?

It's not necessarily the case that there are no better alternatives, but I certainly wouldn't want to be prevented from upgrading to newer versions of Django on existing projects without having to significantly refactor code.

Having said that, personally I see almost no value at all in using CSP or being able to use it without `script-src` `unsafe-inline`. Being strict with CSP disallows all `style` attributes for god's sake. It's ridiculous. For any real-world site you have to allow so much stuff it just adds enormous overhead to every HTTP request with absolutely no benefit.

I also think `json_script` non-script-script-tags-that-must-then-be-JSON.parse'd are very long-winded and convoluted and have disadvantages and no advantages over simply putting the equivalent string in a normal script tag.

> > including when embedded in HTML.
> 
> Not sure about how I feel about that on; embedded in HTML can mean many things. "embedded in a HTML script tag" would be clear as far as the intent is concerned.

The reason I'm avoiding saying that is because of `on...` attributes. I'm not sure "embedded in HTML" can mean anything that would be wrong or misleading?

> > And keeping the example, which is vital as without it there's nothing to tell you that you have to include the quote characters yourself.
> 
> Absolutely, the example is really needed (that said one usually notices during dev that hell break loose if you don't add the quotes -- unless you test solely with numbers or so ;))

Excellent :-)

**Comment by phith0n** [comment]
> Having said that, personally I see almost no value at all in using CSP or being able to use it without script-src unsafe-inline. Being strict with CSP disallows all style attributes for god's sake. It's ridiculous. For any real-world site you have to allow so much stuff it just adds enormous overhead to every HTTP request with absolutely no benefit.
> 
> I also think json_script non-script-script-tags-that-must-then-be-JSON.parse'd are very long-winded and convoluted and have disadvantages and no advantages over simply putting the equivalent string in a normal script tag.

Agree, we must make it possible for both demanding and casual people to develop web applications quickly, no need to follow to a strict design pattern.

`escapejs` is a very simple and useful filter for me.

File: docs/ref/templates/builtins.txt
Comment (nessita) [reply docs/ref/templates/builtins.txt:1647]: Hello everyone! I was trying to fully understand this conversation to try to provide a proposal to unblock this. I was about to reach out to the security team, following the latest comment from @carltongibson, but before doing so I tried to find a way to produce **unsafe** content even when using `escapejs`. I think I found a way, and that this PR should be closed without merging. My example is as follows:

```python
def test_escapejs(self):
    from django.template import Template, Context

    t = Template(
        "{{ var1|escapejs }}\n"
        "<script>{{ var2|escapejs }}</script>\n"
        "<script>{{ var3|escapejs }}</script>\n"
    )
    c = Context({
        "var1": "<script>alert('XSS 1');</script>",
        "var2": 'alert("XSS 2");',
        "var3": "alert(String.fromCharCode(88,83,83))",
    })
    response = t.render(c)
    return HttpResponse(response)
```

The code above would generate:

```html
\u003Cscript\u003Ealert(\u0027XSS 1\u0027)\u003B\u003C/script\u003E
<script>alert(\u0022XSS 2\u0022)\u003B</script>
<script>alert(String.fromCharCode(88,83,83))</script>
```
and the third line is indeed **unsafe**, since the alert is executed. So my advice is to close this PR and do not change the documentation (if anything, discourage the `escapejs` usage or rename it to something that implies "less safeness".

**Review by nessita — CHANGES_REQUESTED** [review]

**Comment by jribbens** [comment]
@nessita I'm afraid you've completely misunderstood this patch and indeed fundamentally what `escapejs` is and does. Your code is insecure whether this patch is accepted or not. It is fundamentally misusing the `escapejs` filter, which has one use and one use only - in some JavaScript code in a template you write `"{{ foo|escapejs }}"` *with the quotes*. My patch makes this more clear, rather than less clear, by adding the example code to show developers how to use it.

Yes, if we could turn back time then I would certainly recommend that this filter output string values with the quotes included, but we can't do that now as it would break backwards compatibility.

This patch is not about fixing the behaviour of the `escapejs` filter, which although not ideal is not broken if used correctly, it is about fixing its documentation, which as it stands is frankly just nonsensical gibberish.

**Comment by nessita** [comment]
@jribbens Thanks for the clarification, I will re-read all the comments with that in mind and re-reply next week.

**Comment by nessita** [comment]
(Just a small clarification: I want to mention that since English is not my native language (and I have also discussed this with other individuals who are in the same situation), when reading the term "JavaScript string," an initial interpretation could be "a string containing JavaScript code" rather than a "text enclosed in quotes." Therefore, I personally believe that providing further clarification about what is meant by a "JavaScript string" could greatly help in understanding.)

**Comment by jribbens** [comment]
> First of all, I would like to kindly ask @jribbens to please tone down their responses.

Sorry, I don't understand what you are referring to here.

> 1- The author of this PR would like the `escapejs` documentation to be changed so it's more explicit about the intended use of the filter. I don't think this is controversial and we all agree it can be improved.
> 
> 2- The author would like the docs to be explicit about the level of "safeness" that the filter guarantees, when used in a specific way. I think this is the controversial bit, since I (and from the comments, other reviewers as well) believe that when Django users read the qualifier "safe", they tend to expect a lot of that and think something along the lines of "safe in almost every context and prevents XSS attacks".

That is indeed what the filter does, when used correctly. Part of the issue here is that the filter used to work a different way, which wasn't "safe" - but then it was completely re-written to make it safe, but when that change happened they forgot to update the documentation. It is now just as safe as any of the other Django filters, e.g. `escape`.

> With that in mind, and 1- using the already suggested paragraphs, and 2- expanding on the bits that (also) caused me confusion (I'm not a JS expert), I would propose that we resort to the "less is more" approach while staying on the secure-by-default side, and using something like:
> 
> > Escapes characters for use inside JavaScript strings (text written inside single or double quotes). Using this filter **only** makes the string safe when used in JavaScript string literals (please see examples below). Note that it does not make the string safe for use in "JavaScript template literals" (i.e. the JavaScript backtick syntax), and any other use is not supported. It is generally recommended however that data should be passed using HTML `data-` attributes, or the `json_script` filter, rather than in embedded JavaScript.
> 
> Then, and following the original proposal, I think we could include examples for both "embed in HTML" cases: using it with `<script>` tag and with the `on` directive.
> 
> Thoughts? Thank you! Natalia.

I generally like your suggestion, but note that the correct usage is more restricted that your words explain. In particular, the following is not going to work properly:
```js
const foo = "lorem ipsum \{{ foo|escapejs }}"
```
The filter *must* be used *only* as literally `"{{` or `'{{` and the value to be output and then `|escapejs }}"` or `|escapejs }}'`. (Hence why I said above it is a great shame they didn't originally make it so that it output the quotes itself.)

> (Just a small clarification: I want to mention that since English is not my native language (and I have also discussed this with other individuals who are in the same situation), when reading the term "JavaScript string," an initial interpretation could be "a string containing JavaScript code" rather than a "text enclosed in quotes." Therefore, I personally believe that providing further clarification about what is meant by a "JavaScript string" could greatly help in understanding.)

Indeed. And also the existing documentation's reference to "JavaScript template literals", which I (and presumably other people) took to mean "JavaScript code which is inside a Django template" rather than what it actually means, "JavaScript backtick syntax".

I've updated the PR to reflect your wording plus my comments above. Unfortunately this required asking git to do something more complicated than just commit something to HEAD, and hence inevitably went horribly wrong and I've had to delete the whole repository and start again from scratch in PR #16897.

**Pull request closed without merging**
